libp2p/network/connection/swarm_connection.py

```diff
@@ -79,28 +78,26 @@ def __init__(self, muxed_conn: IMuxedConn, swarm: "Swarm") -> None:
             self.remove_stream(net_stream)
 
     async def _handle_muxed_stream(self, muxed_stream: IMuxedStream) -> None:
-        net_stream = self._add_stream(muxed_stream)
+        net_stream = await self._add_stream(muxed_stream)
         if self.swarm.common_stream_handler is not None:
-            await self.run_task(self._call_stream_handler(net_stream))
+            await self._call_stream_handler(net_stream)
 
-    def _add_stream(self, muxed_stream: IMuxedStream) -> NetStream:
+    async def _add_stream(self, muxed_stream: IMuxedStream) -> NetStream:
         net_stream = NetStream(muxed_stream)
         self.streams.add(net_stream)
-        self.swarm.notify_opened_stream(net_stream)
+        await self.swarm.notify_opened_stream(net_stream)
         return net_stream
 
-    def _notify_disconnected(self) -> None:
-        self.swarm.notify_disconnected(self)
-
-    async def start(self) -> None:
-        await self.run_task(self._handle_new_streams())
+    async def _notify_disconnected(self) -> None:
+        await self.swarm.notify_disconnected(self)
 
-    async def run_task(self, coro: Awaitable[Any]) -> None:
-        self._tasks.append(asyncio.ensure_future(coro))
+    async def run(self) -> None:
+        self.manager.run_task(self._handle_new_streams)
```

Comment:
I *think* that you could just replace the whole body of this function with `await self._handle_new_streams()`

Reply:
Good catch!